src/main/java/net/starlark/java/eval/StringModule.java

```diff
@@ -526,10 +526,16 @@ public String title(String self) throws EvalException {
   private static int stringFind(boolean forward, String self, String sub, Object start, Object end)
       throws EvalException {
     long indices = substringIndices(self, start, end);
-    // Unfortunately Java forces us to allocate here, even though
-    // String has a private indexOf method that accepts indices.
-    // Fortunately the common case is self[0:n].
-    String substr = self.substring(lo(indices), hi(indices));
+    int startpos = lo(indices);
+    int endpos = hi(indices);
+    // Unfortunately Java forces us to allocate here in the general case, even
+    // though String has a private indexOf method that accepts indices.
+    // The common cases of a search of the full string or a forward search with
+    // a custom start position do not require allocations.
+    if (forward && endpos == self.length()) {
+      return self.indexOf(sub, startpos);
+    }
+    String substr = self.substring(startpos, endpos);
     int subpos = forward ? substr.indexOf(sub) : substr.lastIndexOf(sub);
     return subpos < 0
```

Comment:
Nit: instead of `subpos + lo(indices)` on line 542, we can do `subpos + startpos`.

Reply:
Changed